### PR TITLE
catalog: Add env arg to trigger builtin migration

### PR DESCRIPTION
**Pull request by jkosh44**
This commit adds a CLI argument to environmentd that can trigger builtin migrations. Usually builtin migrations can only be triggered by modifying builtin item definitions that are compiled into the binary, making them extremely difficult to test. The CLI arg added in this commit "tricks" environmentd into thinking that builtin tables are migrated making it much easier to test.

### Motivation
This PR adds a feature that has not yet been specified.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

File: src/environmentd/src/bin/environmentd/main.rs [hunk 1]
```diff
@@ -541,6 +545,24 @@ pub struct Args {
...
+
+    // === Testing options. ===
+    /// Injects arbitrary whitespace into builtin table fingerprints, which can
+    /// trigger builtin item migrations. The amount of whitespace is determined
+    /// by
+    /// `really_dangerous_do_not_call_this_in_production_table_fingerprint_whitespace_version`.
+    /// This argument is meant for testing only and as the names suggests
+    /// should not be set in production.
+    #[clap(long, arg_enum, requires = "unsafe-mode")]
+    really_dangerous_do_not_call_this_in_production_table_fingerprint_whitespace:
```
Comment (benesch) [inline src/environmentd/src/bin/environmentd/main.rs:557]: Can we just call this `unsafe`, for symmetry with the existing `unsafe-mode` option? As fun as really-dangerous-dont-call-this-in-production is I think it's nice to have a single consistent term we use everywhere.

```suggestion
    unsafe_table_fingerprint_whitespace:
```

File: src/environmentd/src/bin/environmentd/main.rs [hunk 2]
```diff
@@ -541,6 +545,24 @@ pub struct Args {
...
+        Option<DangerousTableFingerprintWhitespace>,
+    /// Controls the amount of whitespace injected by
+    /// `really_dangerous_do_not_call_this_in_production_table_fingerprint_whitespace`.
+    /// Incrementing this value can allow triggering multiple builtin
+    /// migrations from a single test. This argument is meant for testing only
+    /// and as the names suggests should not be set in production.
+    #[clap(long, requires = "unsafe-mode", default_value = "1")]
+    really_dangerous_do_not_call_this_in_production_table_fingerprint_whitespace_version: usize,
```
Comment (benesch) [inline src/environmentd/src/bin/environmentd/main.rs:565]: (Same comment as above.)

**Review by benesch — APPROVED** [review]

File: src/environmentd/src/bin/environmentd/main.rs (hunk 1 above)
Comment (jkosh44) [reply src/environmentd/src/bin/environmentd/main.rs]: Done

File: src/environmentd/src/bin/environmentd/main.rs (hunk 2 above)
Comment (jkosh44) [reply src/environmentd/src/bin/environmentd/main.rs]: Done

**Review by benesch — APPROVED** [review]
Still LGTM.

File: src/environmentd/src/bin/environmentd/main.rs [hunk 3]
```diff
@@ -541,6 +545,23 @@ pub struct Args {
     // === Tracing options. ===
     #[clap(flatten)]
     tracing: TracingCliArgs,
+
+    // === Testing options. ===
+    /// Injects arbitrary whitespace into builtin table fingerprints, which can
+    /// trigger builtin item migrations. The amount of whitespace is determined
+    /// by
+    /// `really_dangerous_do_not_call_this_in_production_table_fingerprint_whitespace_version`.
```
Comment (def-) [inline src/environmentd/src/bin/environmentd/main.rs:553]:
```suggestion
    /// `unsafe_builtin_table_fingerprint_whitespace_version`.
```

File: src/environmentd/src/bin/environmentd/main.rs [hunk 4]
```diff
@@ -541,6 +545,23 @@ pub struct Args {
...
+    /// This argument is meant for testing only and as the names suggests
+    /// should not be set in production.
+    #[clap(long, arg_enum, requires = "unsafe-mode")]
+    unsafe_builtin_table_fingerprint_whitespace: Option<UnsafeBuiltinTableFingerprintWhitespace>,
+    /// Controls the amount of whitespace injected by
+    /// `really_dangerous_do_not_call_this_in_production_table_fingerprint_whitespace`.
```
Comment (def-) [inline src/environmentd/src/bin/environmentd/main.rs:559]:
```suggestion
    /// `unsafe_builtin_table_fingerprint_whitespace`.
```

File: src/environmentd/src/bin/environmentd/main.rs [hunk 5]
```diff
@@ -541,6 +545,23 @@ pub struct Args {
...
+    /// Incrementing this value can allow triggering multiple builtin
+    /// migrations from a single test. This argument is meant for testing only
+    /// and as the names suggests should not be set in production.
+    #[clap(long, requires = "unsafe-mode", default_value = "1")]
+    unsafe_builtin_table_fingerprint_whitespace_version: usize,
```
Comment (def-) [inline src/environmentd/src/bin/environmentd/main.rs:564]: This is a bit unergonomic for tests, we have to override and increase the size for each restart in order to stress test, right?

**Review by def- — CHANGES_REQUESTED** [review]

File: src/environmentd/src/bin/environmentd/main.rs (hunk 5 above)
Comment (def-) [reply src/environmentd/src/bin/environmentd/main.rs:564]: Actually I don't see where this variable is used. We appear to only add one "\n" in each case

**Comment by def-** [comment]
I tried using this: https://github.com/MaterializeInc/materialize/pull/28554
It seems testdrive also needs to be adapted:
```
error: catalog inconsistency: catalog state

Caused by:
    failed to read on-disk catalog state: cannot commit a transaction in a read-only catalog

Stack backtrace:
   0: anyhow::__private::format_err
   1: mz_testdrive::run_line_reader::{closure#0}
   2: mz_testdrive::run_string::{closure#0}
   3: testdrive::main::{closure#0}
   4: <tracing::instrument::Instrumented<testdrive::main::{closure#0}> as core::future::future::Future>::poll
   5: <tokio::runtime::park::CachedParkThread>::block_on::<tracing::instrument::Instrumented<testdrive::main::{closure#0}>>
   6: <tokio::runtime::runtime::Runtime>::block_on::<testdrive::main::{closure#0}>
   7: testdrive::main
   8: std::sys_common::backtrace::__rust_begin_short_backtrace::<fn(), ()>
   9: std::rt::lang_start::<()>::{closure#0}
  10: std::rt::lang_start_internal
  11: main
  12: <unknown>
  13: __libc_start_main
  14: <unknown>
```

File: src/environmentd/src/bin/environmentd/main.rs (hunk 5 above)
Comment (jkosh44) [reply src/environmentd/src/bin/environmentd/main.rs:564]: > This is a bit unergonomic for tests, we have to override and increase the size for each restart in order to stress test, right?

Yeah if you want to test multiple builtin item migrations in the same test, then you'll have to increment this for each restart. Is there something that would be more ergonomic?

> Actually I don't see where this variable is used. We appear to only add one "\n" in each case

It's used right below this on line 606

```Rust
let whitespace = "\n".repeat(args.unsafe_builtin_table_fingerprint_whitespace_version);
```

**Comment by jkosh44** [comment]
> I tried using this: #28554 It seems testdrive also needs to be adapted:
> 
> ```
> error: catalog inconsistency: catalog state
> 
> Caused by:
>     failed to read on-disk catalog state: cannot commit a transaction in a read-only catalog

I think the correct default value for `force_migrations` is `None`. We should only set the flag for a select targeted number of tests. I also think that is should not be set in any testdrive test. Testdrive will attempt to open the catalog in read-only mode in a couple of places. If the flag is set and read-only mode thinks that there's unnapplied migrations, then it will fail with the error you're seeing.

File: src/environmentd/src/bin/environmentd/main.rs (hunk 5 above)
Comment (jkosh44) [reply src/environmentd/src/bin/environmentd/main.rs:564]: Another option I considered was to omit `unsafe_builtin_table_fingerprint_whitespace_version` and only allow a single migration per test.

**Comment by def-** [comment]
> I also think that is should not be set in any testdrive test

Most tests use testdrive.

**Comment by jkosh44** [comment]
> > I also think that is should not be set in any testdrive test
> 
> Most tests use testdrive.

Hmm, well as long as `consistency-checks` are disabled then it should be fine. Otherwise we would need to make sure that the consistency checks in testdrive always has matching behavior to what `unsafe_builtin_table_fingerprint_whitespace` and `unsafe_builtin_table_fingerprint_whitespace_version` are set to, which would be difficult.

**Comment by jkosh44** [comment]
@def- let me know if there's a specific change you'd like me to make or if you think the PR should be rejected.

**Comment by def-** [comment]
Currently we have no tests with consistency check disabled. I'll try running the tests in a one-off run with consistency check off: https://buildkite.com/materialize/test/builds/86964
I think there is value in this approach, but I'm not sure yet how to best enable it in our existing tests in a permanent way.

**Comment by jkosh44** [comment]
> I think there is value in this approach, but I'm not sure yet how to best enable it in our existing tests in a permanent way.

I think it would be a mistake to enable this in our existing tests. The idea behind these flags is to use them in a small number of targeted tests meant to stress builtin item migrations specifically. Maybe we could use this flag in our existing upgrade tests but I think we should still run all upgrade tests at least once without this flag.

**Comment by def-** [comment]
I see, we could definitely enable this in an upgrade test.

In the nightly I triggered with this all tests succeeded except [🏎️ testdrive with SIZE 8](https://buildkite.com/materialize/nightly/builds/8842#0190f144-a761-41ca-8dff-fa15989519b0)
```
testdrive-fivetran-destination-1  | thread 'tokio-runtime-worker' panicked at src/fivetran-destination/src/destination/config.rs:184:13: tokio-postgres connection error: error communicating with the server: Connection reset by peer (os error 104)
```
```
fivetran-destination.td:126:1: error: Write Batch Response did not match expected

                    response: WriteBatchResponse {
    response: Some(
        Failure(
            "filesystem error: Custom { kind: Other, error: Error { kind: Closed, cause: None } }: flushing record: replace_files: replace files: write_batch",
        ),
    ),
}

                    expected: WriteBatchResponse {
    response: Some(
        Success(
            true,
        ),
    ),
}
     |
 113 | $ postgres-execute c ... [rest of line truncated for security]
 125 | 
 126 | $ fivetran-destination action=write_batch
     | ^
```

**Review by def- — APPROVED** [review]

**Pull request merged**